new(blog): how we sign and verify plugins and rules (#1182)
falcosecurity/falco-website: content/en/blog/sign-verify-plugins-rules/index.md

On the other side, safeguards with an high utility value can require a considerable amount of effort to be implemented from end-to-end (producer and consumer sides).

MITRE and OpenSSF increased their efforts to improve SSC security recently in general and in the open source, by providing frameworks and tools that increase the utility to cost ratio.
MITRE for example proposes an end-to-end framework to preserve the integrity of the software supply chain. OpenSSF develops the SLSA framework that groups several security best-practices for open source projects.

Moreover, the in-toto Attestation Framework defines a fixed, lightweight Statement that communicates some information about the execution of the software supply chain, such as if the source was reviewed, or if it went through a SLSA conformant build process.

> Supply chain security is important to be constantly increased through safeguards placed as soon as possible in the supply chain ("shift left") and needs to be maintained in a continuous manner reducing the operational burden (DevSecOps) as technologies evolve and so attacks do.

Reviewer: WDYT @LucaGuerra?

```suggestion
Supply chain security is important to be constantly increased through safeguards placed as soon as possible in the supply chain ("shift left") and needs to be maintained in a continuous manner reducing the operational burden (DevSecOps) because technologies evolve and so attacks do.
```

{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "size": 242,
    "digest": "sha256:ef254edd0dd0ae2b9b6ef0471c915682c0ba31858fe45bc817a951cad715fe8d"
  },
  "layers": [
    {
      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
      "size": 263,
      "digest": "sha256:f8aae8441b2d41419ae99d4979d6eab1543d81a269e4721fdbff0d837571aa4d",
      "annotations": {
        "dev.cosignproject.cosign/signature": "MEUCIGvgIIu9JX5vVCZ5YHsoVe3wpVoIianJvzkCZUIClVlAAiEAyx7GZztNDzCJN5Fjg8vwq0pv7PH3TrseIXN1ou35Jmw=",
        "dev.sigstore.cosign/bundle": "{\"SignedEntryTimestamp\":\"MEYCIQDsAPhTbHc7w+LHpipbx0HSEIlmjjbAszF5ChHRO1jyKgIhAJrDmeLtzO72fGD1WHabY8Mwz7Qph9CzDg+EMkC9JR7A\",\"Payload\":{\"body\":REDACTED}}",
        "dev.sigstore.cosign/certificate": "-----BEGIN CERTIFICATE-----\nREDACTED\n-----END CERTIFICATE-----\n",
        "dev.sigstore.cosign/chain": "-----BEGIN CERTIFICATE-----\nREDACTED=\n-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----\nREDACTED\n-----END CERTIFICATE-----"
      }
    }
  ]
}
```

> and see that it contains a layer of type `application/vnd.dev.cosign.simplesigning.v1+json`. Also, its annotation metadata helps cosign to authenticate the signature against the Sigstore's [Rekor transparency log](https://docs.sigstore.dev/logging/overview/).

Reviewer: ```suggestion
You can see that it contains a layer of type `application/vnd.dev.cosign.simplesigning.v1+json`. Also, its annotation metadata helps cosign to authenticate the signature against the Sigstore's [Rekor transparency log](https://docs.sigstore.dev/logging/overview/).
```

---

> title: "How we Sign and Verify Plugins and Rules"

Reviewer 2: Overall comment: I think you need an introductory paragraph that describes what this blog is about and why folks should care. 

I think you could move the general information on SSC attacks up here to the top to set context for the blog. 

Software supply chain attacks are on the rise, especially among open source projects. Over the past few years we've seen an increasing number of attacks, offering proof that attackers find this a viable attack vector. 

To help thwart these threats inside of Falco's own software supply chain we've introduced new processes to sign and verify Falco plugins. In this blog we'll walk you through how we implemented those changes. Our hope is by sharing information about our journey, you can implement similar changes to help secure your own software supply chains. 

Something like that.

linktitle: "How we Sign and Verify Plugins and Rules"
date: 2023-10-12
author: Massimiliano Giovagnoli, Luca Guerra
slug: sign-verify-plugins-rules
images:
  - /blog/how-we-sign-and-verify-plugins-and-rules/images/falcoctl-signature-featured.png
---

## Falco v0.36.0 and the SSC security

> The latest stable v0.36.0, alongside falcoctl 0.6.1 and the 0.7.0 Helm chart brought new features and new improvements in the supply chain security of Falco artifacts. The two main artifacts are plugins and rule sets as you already know. They're shipped in the OCI specification format and distributed through the official Falcosecurity OCI repositories.

Reviewer 2: "new improvements in the supply chain security of Falco artifacts" -> improvements to the security of Falco's software supply chain artifacts. 

You could leave out "software" but "software supply chain" is something folks would likely search on.

> The latest stable v0.36.0, alongside falcoctl 0.6.1 and the 0.7.0 Helm chart brought new features and new improvements in the supply chain security of Falco artifacts. The two main artifacts are plugins and rule sets as you already know. They're shipped in the OCI specification format and distributed through the official Falcosecurity OCI repositories.

Reviewer 2: "The two main artifacts are plugins and rule sets as you already know" I tend to avoid phrases like "you already know" because if the reader doesn't already know they may stop reading or they may feel like they don't know enough about the topic or it just makes them feel bad. 

Also The two many artifacts of what?  "Falco's two main artifacts are plugins and rules.

> Software supply chain attacks aim at injecting malicious code into software components like the above, to compromise downstream users. These types of attacks are among the primary threats in today's threat landscape.

Reviewer 2: I don't think you need "like the above".

> Software supply chain attacks aim at injecting malicious code into software components like the above, to compromise downstream users. These type of attacks are among the primary threats in the today's thread landscape.

Reviewer: ```suggestion
Software supply chain attacks aim at injecting malicious code into software components like the above, to compromise downstream users. These types of attacks are among the primary threats in the today's thread landscape.
```

> Software supply chain attacks aim at injecting malicious code into software components like the above, to compromise downstream users. These type of attacks are among the primary threats in the today's thread landscape.

Reviewer: ```suggestion
Software supply chain attacks aim at injecting malicious code into software components like the above, to compromise downstream users. These type of attacks are among the primary threats in today's thread landscape.
```

In particular, attackers abuse trust relationships existing between the different open-source stakeholders. The increase in attacks on open-source software throughout the last few years demonstrates that attackers consider them a viable means for spreading malware.

> ## SSC safeguards

Reviewer 2: I think you need a quick introduction for this section.  

for example

the task of securing your software supply chain may seem daunting at first glance. But there are a number of ways you can help ease this effort.

Reviewer: Thank you @mikegcoleman!

Reviewer: ```suggestion
## SSC safeguards

Securing the software supply chain may seem daunting at first glance, but there are a lot of safeguards that can be put in action. And there are ways to categorize them, and ways to prioritize them.
```

> Safeguards against supply chain attacks can be classified by control type: directive, preventive, detective, corrective, and recovery. But there ain't no such thing as a free lunch. Besides safeguards classification, the utility-to-cost ratio can also be an important factor in deciding where to start in improving the supply chain security of software, and can be pretty easy to measure it.

Reviewer 2: "safeguards classification" -> "safeguard classifications"

I'd always say software supply chain. I imagine it's implied to most folks, but I think it is better for SEO.

Reviewer 3: ```suggestion
Safeguards against supply chain attacks can be classified by control type: directive, preventive, detective, corrective, and recovery. But there ain't no such thing as a free lunch. Besides safeguard classifications, the utility-to-cost ratio can also be an important factor in deciding where to start in improving the supply chain security of software, and can be pretty easy to measure it.
```

> Safeguards against supply chain attacks can be classified by control type: directive, preventive, detective, corrective and recovery. But there ain't no such thing as a free lunch. Besides safeguards classification, the utility to cost ratio can also be an important factor to decide where to start from improving the supply chain security of a software, and can be pretty easy to measure it.

Reviewer: ```suggestion
Safeguards against supply chain attacks can be classified by control type: directive, preventive, detective, corrective, and recovery. But there ain't no such thing as a free lunch. Besides safeguards classification, the utility-to-cost ratio can also be an important factor in deciding where to start in improving the supply chain security of software, and can be pretty easy to measure it.
```

> There are cheap preventive safeguards that can be implemented in open source projects especially, where stakeholders platea can be pretty wide considering the contributions. For example branch protection rules are usually simple per-code repository configurations in providers (e.g. GitHub) and alongside pull-request based flows enforcing code review quorum, are also standard best practices nowadays. The same applies to reproducible builds, dependency pinning, build steps isolation, MFA authentication to repository providers, etc.

Reviewer: ```suggestion
There are cheap preventive safeguards that can be implemented in open source projects especially, where stakeholders' plateau can be pretty wide considering the contributions. For example, branch protection rules are usually simple per-code repository configurations in providers (e.g. GitHub) and alongside pull request-based flows enforcing code review quorum, are also standard best practices nowadays. The same applies to reproducible builds, dependency pinning, build steps isolation, MFA authentication to repository providers, etc.
```

Author: I think `stakeholders platea` was right.

Reviewer: Thanks @LucaGuerra , I updated the suggestion

Author: I updated the text.

> On the other side, safeguards with an high utility value can require a considerable amount of effort to be implemented end-to-end (producer and consumer sides).

Reviewer: ```suggestion
On the other side, safeguards with a high utility value can require a considerable amount of effort to be implemented end-to-end (producer and consumer sides).
```

> MITRE and OpenSSF increased their efforts to improve SSC security recently in general and in the open source, by providing frameworks and tools that increase the utility-to-cost ratio.

Reviewer 2: You use SSC here as an abbreviation, which is good. But, you should go back to very first time you use software supply chain and put (SSC) behind it - "software supply chain (SSC)"

then you can replace every other instance of "software supply chain" with SSC if you want.

Author: For this I think replacing it in the first heading that mentions it would be good enough

Reviewer: I agree, thank you @mikegcoleman

> MITRE and OpenSSF increased their efforts to improve SSC security recently in general and in the open source, by providing frameworks and tools that increase the utility to cost ratio.

Reviewer: ```suggestion
MITRE and OpenSSF increased their efforts to improve SSC security recently in general and in the open source, by providing frameworks and tools that increase the utility-to-cost ratio.
```

> MITRE for example proposes an end-to-end framework to preserve the integrity of the software supply chain. OpenSSF develops the SLSA framework that groups several security best practices for open-source projects.

Reviewer 2: Might be nice to link to articles about the framework and a "What is SLSA" blog or something like that for readers who need more context

Author: @maxgio92 :point_up: can you help with the links?

Reviewer: Thanks @mikegcoleman, sure @LucaGuerra

Reviewer: Proposed

> MITRE for example proposes an end-to-end framework to preserve the integrity of the software supply chain. OpenSSF develops the SLSA framework that groups several security best-practices for open source projects.

Reviewer: ```suggestion
MITRE for example proposes an end-to-end framework to preserve the integrity of the software supply chain. OpenSSF develops the SLSA framework that groups several security best practices for open-source projects.
```

Moreover, the in-toto Attestation Framework defines a fixed, lightweight Statement that communicates some information about the execution of the software supply chain, such as if the source was reviewed, or if it went through a SLSA conformant build process.

> Supply chain security is important to be constantly increased through safeguards placed as soon as possible in the supply chain ("shift left") and needs to be maintained in a continuous manner reducing the operational burden (DevSecOps) because technologies evolve so attacks do.

Reviewer 2: Supply chain security is important to be constantly increased through safeguards placed as soon as possible in the supply chain ("shift left") and needs to be maintained in a continuous manner reducing the operational burden (DevSecOps) because technologies evolve so attacks do. 

I might rewrite this a bit

"Because new attacks are being introduced almost every day, It's important to be constantly evolving your SSC security tools and processes. These safeguards should be placed as early as possible in the SSC ("shift left"). Additionally, they need to be continuously maintained in order to keep pace with attacker's efforts.

Reviewer: Thank you @mikegcoleman! I like it, it sounds a lot better

Reviewer: ```suggestion
Because new attacks are being introduced almost every day, It's important to be constantly evolving your SSC security tools and processes. These safeguards should be placed as early as possible in the SSC ("shift left"). Additionally, they need to be continuously maintained in order to keep pace with attacker's efforts.
```

> Software artifact signatures are one of the safeguards with a reasonable amount of effort nowadays, thanks also to important projects like Sigstore's [cosign](https://github.com/sigstore/cosign) for artifacts distributed as container images. So this was one of the goals decided for the Falco Software Supply Chain Working Group.

Reviewer 2: "with a reasonable amount of effort' -> with "with good adoption" or "with good momentum"

I think you can drop "for artifacts distributed as container images"

"thanks to important projects like Sigstore's cosign (link), software artifact signatures are one of the SSC safeguards with a good amount of traction. Due to these factors, the Falco Software Supply Chain Working group decided to make implementing these signatures a goal for the Falco project"

Reviewer: Thanks @mikegcoleman! The only thing I'd change is that I'd keep the utility-to-cost ratio as a metric for this type of safeguard instead of the traction or the adoption level of community projects that ease the process.

```suggestion
Thanks to important projects like Sigstore's [cosign](https://github.com/sigstore/cosign), software artifact signatures are one of the SSC safeguards with a high utility-to-cost ratio. Due to these factors, the Falco Software Supply Chain Working group decided to make implementing these signatures a goal for the Falco project"
```

> Software artifacts signatures are one of the safeguards with a reasonable amount of effort nowadays, thanks also to important projects like Sigstore's [cosign](https://github.com/sigstore/cosign) for artifacts distributed as container images. So this was one of the goals decided for the Falco Software Supply Chain Working Group.

Reviewer: ```suggestion
Software artifact signatures are one of the safeguards with a reasonable amount of effort nowadays, thanks also to important projects like Sigstore's [cosign](https://github.com/sigstore/cosign) for artifacts distributed as container images. So this was one of the goals decided for the Falco Software Supply Chain Working Group.
```

## Artifacts distributed as OCI

> Since January 2023 we started distributing Falco plugins and rules as [OCI artifacts](https://github.com/oras-project/artifacts-spec), in addition to other ways. If you are not familiar with the concept, think about OCI artifacts as an unit of data that can be tagged, stored and indexed in container registries alongside your container images but can hold any type of file. You can find the source and the data for Falco's artifacts in the [plugins](https://github.com/falcosecurity/plugins) and [rules](https://github.com/falcosecurity/rules) repositories respectively. We are using GitHub Container Registry to publish them so you can find them in the "Package" section of each repository.

Reviewer: ```suggestion
Since January 2023 we started distributing Falco plugins and rules as [OCI artifacts](https://github.com/oras-project/artifacts-spec), in addition to other ways. If you are not familiar with the concept, think about OCI artifacts as a content-addressable unit of data that can be tagged, stored and indexed in container registries alongside your container images but can hold any type of file. You can find the source and the data for Falco's artifacts in the [plugins](https://github.com/falcosecurity/plugins) and [rules](https://github.com/falcosecurity/rules) repositories respectively. We are using GitHub Container Registry to publish them so you can find them in the "Package" section of each repository.
```

> Since January 2023 we started distributing Falco plugins and rules as [OCI artifacts](https://github.com/oras-project/artifacts-spec), in addition to other ways. If you are not familiar with the concept, think about OCI artifacts as a content-addressable unit of data that can be tagged, stored and indexed in container registries alongside your container images but can hold any type of file. You can find the source and the data for Falco's artifacts in the [plugins](https://github.com/falcosecurity/plugins) and [rules](https://github.com/falcosecurity/rules) repositories respectively. We are using GitHub Container Registry to publish them so you can find them in the "Package" section of each repository.

Reviewer 2: you could drop "in addition to other ways"

Reviewer: ```suggestion
Since January 2023 we started distributing Falco plugins and rules as [OCI artifacts](https://github.com/oras-project/artifacts-spec). If you are not familiar with the concept, think about OCI artifacts as a content-addressable unit of data that can be tagged, stored and indexed in container registries alongside your container images but can hold any type of file. You can find the source and the data for Falco's artifacts in the [plugins](https://github.com/falcosecurity/plugins) and [rules](https://github.com/falcosecurity/rules) repositories respectively. We are using GitHub Container Registry to publish them so you can find them in the "Package" section of each repository.
```

> This brings several advantages, such as:

Reviewer: ```suggestion
Packages Falco's rules and plugins as container artifacts brings several advantages, such as:
```

> * We can create multiple tags, both fixed and floating, for each artifact. Keeping track of the `latest` version is easy as well as the latest version for each major and minor release.

Reviewer 2: "This brings" using "this" here is a bit ambiguous. You might want to be more explicit "packaging Falco's rules and plugins as OCI artifacts brings"

Reviewer: Thank you @mikegcoleman, I totally agree

> * Distribution is handled with a standard protocol implemented by many different container registries, which means that creating [your own rules registry](https://falco.org/blog/gitops-your-falco-rules/), either public or private, is easy and you can use your existing container distribution infrastructure (or any ready to use registry service as offered by all major cloud providers) without installing and maintaining any custom server.

Reviewer: ```suggestion
* Distribution is handled with a standard protocol implemented by many different container registries, which means that creating [your own rules registry](https://falco.org/blog/gitops-your-falco-rules/), either public or private, is easy and you can use your existing container distribution infrastructure (or any ready-to-use registry service as offered by all major cloud providers) without installing and maintaining any custom server.
```

> * There are standard and developing mechanisms and specifications to store artifact metadata, like [link between artifacts](https://opencontainers.org/posts/blog/2023-07-07-summary-of-upcoming-changes-in-oci-image-and-distribution-specs-v-1-1/#2-new-manifest-field-for-establishing-relationships) for signature or SBOM references, so we do not need to re-invent the wheel.

Reviewer: ```suggestion
* There are standard and developing mechanisms and specifications to store artifact metadata, like [link between artifacts](https://opencontainers.org/posts/blog/2023-07-07-summary-of-upcoming-changes-in-oci-image-and-distribution-specs-v-1-1/#2-new-manifest-field-for-establishing-relationships) for signature or SBOM references, so we do not need to reinvent the wheel.
```

> * There are standard and developing mechanisms and specifications to store artifact metadata, like [link between artifacts](https://opencontainers.org/posts/blog/2023-07-07-summary-of-upcoming-changes-in-oci-image-and-distribution-specs-v-1-1/#2-new-manifest-field-for-establishing-relationships) for signature or SBOM references, so we do not need to reinvent the wheel.

Reviewer 2: With both SBOM here and SLSA above - either link to something that that explains what they are or use their full name and put the acronym in parenthesis if you're going to use it again.

> * There are standard and developing mechanisms and specifications to store artifact metadata, like [link between artifacts](https://opencontainers.org/posts/blog/2023-07-07-summary-of-upcoming-changes-in-oci-image-and-distribution-specs-v-1-1/#2-new-manifest-field-for-establishing-relationships) for signature or SBOM references, so we do not need to reinvent the wheel.

Reviewer: ```suggestion
* There are standard and developing mechanisms and specifications to store artifact metadata, like [link between artifacts](https://opencontainers.org/posts/blog/2023-07-07-summary-of-upcoming-changes-in-oci-image-and-distribution-specs-v-1-1/#2-new-manifest-field-for-establishing-relationships) for references to signature or Software Bill Of Materials (SBOM), so we do not need to reinvent the wheel.
```

In [falcoctl](https://github.com/falcosecurity/falcoctl) we have used the CNCF [ORAS](https://oras.land/) library to manage downloading and uploading artifacts to the container registry.

## Cosign keyless signing and verification

> Another very interesting thing you can do with OCI artifacts is now signing them thanks to [cosign](https://github.com/sigstore/cosign)! Cosign supports a very interesting mechanism called [keyless signing](https://edu.chainguard.dev/open-source/sigstore/cosign/an-introduction-to-cosign/#keyless-signing). This allows us to perform artifact signature and verification without the need for creating, maintaining and securing long-lived and opaque signing keys.

Reviewer: ```suggestion
Another very interesting thing you can do with OCI artifacts is now sign them thanks to [cosign](https://github.com/sigstore/cosign)! Cosign supports a very interesting mechanism called [keyless signing](https://edu.chainguard.dev/open-source/sigstore/cosign/an-introduction-to-cosign/#keyless-signing). This allows us to perform artifact signature and verification without the need for creating, maintaining and securing long-lived and opaque signing keys.
```

> Moreover, cosign [supports OCI v1.1](https://www.chainguard.dev/unchained/building-towards-oci-v1-1-support-in-cosign) being able to link security artifacts through OCI references, when supported by the registry.

Reviewer: ```suggestion
Moreover, cosign [supports OCI v1.1](https://www.chainguard.dev/unchained/building-towards-oci-v1-1-support-in-cosign) being able to link security artifacts through OCI references when supported by the registry.
```

## Implementation

### Release pipelines for signed plugins and rules

![](images/falcoctl_signature_pipeline_part.png)

Plugins and rules that we sign are packaged as OCI artifacts.

As for the diagram above, there are two distinct distribution pipelines: one is to update the general Falcoctl index and one is to distribute the single artifact, whether plugin or rule set, to the OCI registry.

#### Container registry update

> At the same time, in the release pipeline, an OCI artifact [is built](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/.github/workflows/upload-oci-artifacts.yaml#L31) with a specific [build tool](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/build/registry/pkg/oci/oci.go#L398) for each artifact updated, and [a signature is generated](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/.github/workflows/upload-oci-artifacts.yaml#L72) and distributed to the registry for each one. The signature is generated with the help of Cosign using the keyless mode, that leverages OIDC with GitHub as the identity provider.

Reviewer: ```suggestion
At the same time, in the release pipeline, an OCI artifact [is built](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/.github/workflows/upload-oci-artifacts.yaml#L31) with a specific [build tool](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/build/registry/pkg/oci/oci.go#L398) for each artifact updated, and [a signature is generated](https://github.com/falcosecurity/plugins/blob/81ccd91dfa2565a99f8e392d50b642aed89799e7/.github/workflows/upload-oci-artifacts.yaml#L72) and distributed to the registry for each one. The signature is generated with the help of Cosign using the keyless mode, which leverages OIDC with GitHub as the identity provider.
```

  },
  "layers": [
    {
      "mediaType": "application/vnd.dev.cosign.simplesigning.v1+json",
      "size": 263,
      "digest": "sha256:f8aae8441b2d41419ae99d4979d6eab1543d81a269e4721fdbff0d837571aa4d",
      "annotations": {
        "dev.cosignproject.cosign/signature": "MEUCIGvgIIu9JX5vVCZ5YHsoVe3wpVoIianJvzkCZUIClVlAAiEAyx7GZztNDzCJN5Fjg8vwq0pv7PH3TrseIXN1ou35Jmw=",
        "dev.sigstore.cosign/bundle": "{\"SignedEntryTimestamp\":\"MEYCIQDsAPhTbHc7w+LHpipbx0HSEIlmjjbAszF5ChHRO1jyKgIhAJrDmeLtzO72fGD1WHabY8Mwz7Qph9CzDg+EMkC9JR7A\",\"Payload\":{\"body\":REDACTED}}",
        "dev.sigstore.cosign/certificate": "-----BEGIN CERTIFICATE-----\nREDACTED\n-----END CERTIFICATE-----\n",
        "dev.sigstore.cosign/chain": "-----BEGIN CERTIFICATE-----\nREDACTED=\n-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----\nREDACTED\n-----END CERTIFICATE-----"
      }
    }
  ]
}
```
You can see that it contains a layer of type `application/vnd.dev.cosign.simplesigning.v1+json`. Also, its annotation metadata helps cosign to authenticate the signature against the Sigstore's [Rekor transparency log](https://docs.sigstore.dev/logging/overview/).

#### Index update

> Each artifact type has its own [registry metadata file](https://github.com/falcosecurity/plugins/blob/master/registry.yaml). The general [falcoctl index.yaml](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml), that refers to all artifact types, [is updated](https://github.com/falcosecurity/test-infra/blob/bc4db5255ad6f85ee437368bc37ee752f980182d/config/jobs/build-plugins/build-plugins.yaml#L2) when a modification to the specific artifact type metadata occurs. In order to verify a keyless signature we need to know in advance some additional metadata about the type of signature and how to verify its authenticity. We are adding this data to the registry metadata so that it can be published to the index by the pipeline. We will see this content in the next section.

Reviewer: ```suggestion
Each artifact type has its own [registry metadata file](https://github.com/falcosecurity/plugins/blob/master/registry.yaml). The general [falcoctl index.yaml](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml), which refers to all artifact types, [is updated](https://github.com/falcosecurity/test-infra/blob/bc4db5255ad6f85ee437368bc37ee752f980182d/config/jobs/build-plugins/build-plugins.yaml#L2) when a modification to the specific artifact type metadata occurs. In order to verify a keyless signature we need to know in advance some additional metadata about the type of signature and how to verify its authenticity. We are adding this data to the registry metadata so that it can be published to the index by the pipeline. We will see this content in the next section.
```

### Transparent signature verification in `falcoctl`

![](images/falcoctl_signature_falco_part.png)

> We have seen that every artifact has now signature data available. This is nice, but to complete this feature we need to make sure that the signature data is automatically checked when we download an artifact. In the same way package signatures are verified when you run `apt-get install`, falcoctl needs to do the same to the OCI artifacts it downloads.

Reviewer: ```suggestion
We have seen that every artifact has now signature data available. This is nice, but to complete this feature we need to make sure that the signature data is automatically checked when we download an artifact. In the same way, package signatures are verified when you run `apt-get install`, falcoctl needs to do the same to the OCI artifacts it downloads.
```

> We have seen that every artifact has now signature data available. This is nice, but to complete this feature we need to make sure that the signature data is automatically checked when we download an artifact. In the same way, package signatures are verified when you run `apt-get install`, falcoctl needs to do the same to the OCI artifacts it downloads.

Reviewer 2: "has now" should be "now has"

> When designing this feature we decided not to change the existing usage of `falcoctl`. You know how when you decide to install a new repository for `apt` you need to also import its public key? Thanks to keyless signing we can avoid this step and make verification transparent. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.

Reviewer: ```suggestion
When designing this feature we decided not to change the existing usage of `falcoctl`. You know how when you decide to install a new repository for `apt` you need to also import its public key? Thanks to keyless signing trusting the Sigstore [Fulcio](https://docs.sigstore.dev/certificate_authority/overview/) CA we can avoid this step and make verification transparent. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.
```

> When designing this feature we decided not to change the existing usage of `falcoctl`. You know how when you decide to install a new repository for `apt` you need to also import its public key? Thanks to keyless signing we can avoid this step and make verification transparent. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.

Reviewer: ```suggestion
When designing this feature we decided not to change the existing usage of `falcoctl`. You know how when you decide to install a new repository for `apt` you need to also import its public key? Thanks to keyless signing we can avoid this step and make verification transparent. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance, the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.
```

> When designing this feature we decided not to change the existing usage of `falcoctl`. You know how when you decide to install a new repository for `apt` you need to also import its public key? Thanks to keyless signing trusting the Sigstore [Fulcio](https://docs.sigstore.dev/certificate_authority/overview/) CA we can avoid this step and make verification transparent. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.

Reviewer 2: "You know how" maybe say "You may have experienced that when . . ."

Again maybe they don't know, so using "you may" includes folks who have not seen it before.

Reviewer: I got it @mikegcoleman, and I agree. Thank you

Reviewer: I've also rephrased a bit the first part of the paragraph:

```suggestion
When designing this feature we decided not to change the existing usage of `falcoctl`. For example, you may already know that when you decide to install a new `apt` repository you need to also import its public key. When installing Falco artifacts with `falcoctl`, we can avoid this step and make verification transparent, thanks to keyless signing by trusting the Sigstore [Fulcio](https://docs.sigstore.dev/certificate_authority/overview/) CA. In the section [above](#Index-update) we introduced the concept of index files. Every time you use falcoctl, there is an index file that tells details about artifacts distributed by the Falco organization. For instance, the `cloudtrail` plugin has [an entry](https://github.com/falcosecurity/falcoctl/blob/gh-pages/index.yaml#L15) with several pieces of metadata, including the OCI artifact coordinates: `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`. Of course, anyone can distribute and [install their own index file](https://falco.org/blog/falcoctl-install-manage-rules-plugins/#index) for their plugins and rules.
```

Upon request to install the artifact falcoctl is going to check the index file, resolve the name, identify the digest from the registry, pull the signature and validate it by using cosign as a library. Only once the signature is verified, falcoctl will allow the installation of the plugin with that specific digest. If the signature data does not match the installation will be halted with an error.

You can try it yourself to make sure that the signature works!

```
sudo docker run -it --entrypoint sh falcosecurity/falcoctl:0.6.1
# falcoctl artifact install cloudtrail:0.9 --rulesfiles-dir=/tmp --plugins-dir=/tmp
```

![](https://hackmd.io/_uploads/Byldjgnla.png)

But there is more! If you have `cosign` installed you can independently verify the signature without using falcoctl or any index file:

```
$ cosign verify ghcr.io/falcosecurity/plugins/plugin/cloudtrail:0.9.0 --certificate-oidc-issuer=https://token.actions.githubusercontent.com --certificate-identity-regexp=https://github.com/falcosecurity/plugins/

Verification for ghcr.io/falcosecurity/plugins/plugin/cloudtrail:0.9.0 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated

>   - Existence of the claims in the transparency log was verified offline

Reviewer: ```suggestion
  - The existence of the claims in the transparency log was verified offline
```

Author: This is cosign's output :)

  - The code-signing certificate was verified using trusted certificate authority certificates

[...]
```

> This is one of the main advantages of relying on already existing and popular container artifacts signing technologies of the cloud-native world. Authentication of the artifacts is decoupled from the producer, and consumers can always choose the method they prefer.

Reviewer 2: Try this for the first sentence. 

This is one of the main advantages of relying on widely adopted cloud native contain artifact signing technologies.

Reviewer: I like it @mikegcoleman, thank you a lot for this review!

Reviewer: ```suggestion
This is one of the main advantages of relying on widely adopted container artifacts signing technologies of the cloud-native world. Authentication of the artifacts is decoupled from the producer, and consumers can always choose the method they prefer.
```

### Security Analysis

> With the above scheme we made steps forward and implemented best practices towards better supply chain security in the Falco ecosystem. We would like to conclude our discussion with the main takeaways for a security practitioner.

Reviewer: ```suggestion
With the above scheme, we made steps forward and implemented best practices towards better supply chain security in the Falco ecosystem. We would like to conclude our discussion with the main takeaways for a security practitioner.
```

> With the above scheme, we made steps forward and implemented best practices towards better supply chain security in the Falco ecosystem. We would like to conclude our discussion with the main takeaways for a security practitioner.

Reviewer 2: software supply chain or just use SSC

#### Safeguards against registry tampering, secure by default

> Making use of container registries to distribute artifacts is convenient, and allows for easy replication of the data across multiple regions or providers. But even if one of those registries is at risk of being compromised, like it [happened to AWS ECR in December 2022](https://www.bleepingcomputer.com/news/security/amazon-ecr-public-gallery-flaw-could-have-wiped-or-poisoned-any-image/), clients will be protected from potentially malicious applications. If **an artifact is replaced in the registry** without going through the appropriate approval and open source release pipeline, **clients will identify that the signature does not match and automatically refuse to install** any compromised package. This is only possible because cosign signature information is distributed in a file which is separate from the artifact itself. The feature is on by default and requires no extra action from users.

Reviewer: ```suggestion
Making use of container registries to distribute artifacts is convenient, and allows for easy replication of the data across multiple regions or providers. But even if one of those registries is at risk of being compromised, like it [happened to AWS ECR in December 2022](https://www.bleepingcomputer.com/news/security/amazon-ecr-public-gallery-flaw-could-have-wiped-or-poisoned-any-image/), clients will be protected from potentially malicious applications. If **an artifact is replaced in the registry** without going through the appropriate approval and open source release pipeline, **clients will identify that the signature does not match and automatically refuse to install** any compromised package. This is only possible because cosign signature information is distributed in a file that is separate from the artifact itself. The feature is on by default and requires no extra action from users.
```

> Making use of container registries to distribute artifacts is convenient, and allows for easy replication of the data across multiple regions or providers. But even if one of those registries is at risk of being compromised, like it [happened to AWS ECR in December 2022](https://www.bleepingcomputer.com/news/security/amazon-ecr-public-gallery-flaw-could-have-wiped-or-poisoned-any-image/), clients will be protected from potentially malicious applications. If **an artifact is replaced in the registry** without going through the appropriate approval and open source release pipeline, **clients will identify that the signature does not match and automatically refuse to install** any compromised package. This is only possible because cosign signature information is distributed in a file that is separate from the artifact itself. The feature is on by default and requires no extra action from users.

Reviewer 2: maybe just use "risk of being compromised" as the anchor for the link and delete "like it happened to AWS ECR in December 2022"

Reviewer 2: "With Falco 0.36.1 this feature is on by default for rules and plugins, requiring no extra action from users."

#### Independent signature verification

It is now possible for **all Falco rules and plugins users to independently verify the authenticity** of every artifact they download. We implemented this feature staying true to the open nature of Falco, so by design, you do not need any Falco-specific tool (like `falcoctl`) to verify your artifact and you can do it manually with `cosign` to implement even tighter controls.

> #### Attack surface

Reviewer 2: This feels kind of weird being at the end, maybe move it up above as a preface to why it's critical to secure the software supply chain. 

I think the paragraph above makes much better ending.

Author: This paragraph describes the current attack surface, which still exists of course after our improvements but it is reduced.

Reviewer: Maybe we could rename the section with "Conclusion and take-aways"

> As security practitioners we should always analyze the attack surface of the security schemes we design and update our threat model accordingly. In this case we can identify that:

Reviewer: ```suggestion
As security practitioners, we should always analyze the attack surface of the security schemes we design and update our threat model accordingly. In this case we can identify that:
```

> * If the source repositories are compromised, so will be the packages. Of course, the open source repositories act as the single source of truth for our plugins and rules. If an attacker manages to merge malicious code and release a malicious plugin they will be able to compromise users once they update or install a new version. Release pipelines are the only component that can generate a proper signature for the the artifacts.

Reviewer: ```suggestion
* If the source repositories are compromised, so will the packages. Of course, the open-source repositories act as the single source of truth for our plugins and rules. If an attacker manages to merge malicious code and release a malicious plugin they will be able to compromise users once they update or install a new version. Release pipelines are the only component that can generate a proper signature for the artifacts.
```

* If the index file is compromised, so will the packages. The index file maps artifact names (e.g. `cloudtrail`) to their location in the container registry (e.g. `ghcr.io/falcosecurity/plugins/plugin/cloudtrail`). If that entry is modified to point to an attacker-controlled registry, possibly alongside an attacker-controlled signature, clients will download a potentially malicious artifact instead of the official one.
* At this point in time, the entire infrastructure is hosted by GitHub, making it a potential single point of failure and an entity that the Falco project completely trusts. However, even if parts of the infrastructure (container registry, CDN for index distribution, ...) were not, GitHub release pipelines would still have permissions or access tokens to operate those locations.

> 

Reviewer: Also, we could conclude the blog with something like this:

```suggestion
## Le'ts in touch

Given the above considerations, this is a work in progress toward a more secure supply chain for Falco and its artifacts.

We'll keep you up to date with the next steps and remember, if you want to be involved you're more than welcome, don't hesitate to reach out to the Falco Software Supply Chain Security Working Group in the [#falco-sscs-wg](https://cloud-native.slack.com/archives/C04ET5YR014) Slack channel in the CNCF workspace.
```

Author: Done!
